processor/maskprocessor/processor.go

```diff
@@ -0,0 +1,191 @@
+// Copyright  observIQ, Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package maskprocessor
+
+import (
+	"context"
+	"fmt"
+	"regexp"
+
+	"go.opentelemetry.io/collector/component"
+	"go.opentelemetry.io/collector/pdata/pcommon"
+	"go.opentelemetry.io/collector/pdata/plog"
+	"go.opentelemetry.io/collector/pdata/pmetric"
+	"go.opentelemetry.io/collector/pdata/ptrace"
+	"go.uber.org/zap"
+)
+
+const (
+	resourceField   = "resource"
+	attributesField = "attributes"
+	bodyField       = "body"
+)
+
+// processor is the processor used to mask data.
+type processor struct {
+	logger *zap.Logger
+	cfg    *Config
+	rules  map[string]*regexp.Regexp
+}
+
+// newProcessor creates a new mask processor.
+func newProcessor(logger *zap.Logger, cfg *Config) *processor {
+	return &processor{
+		logger: logger,
+		cfg:    cfg,
+	}
+}
+
+// start is used to start the processor.
+func (p *processor) start(context.Context, component.Host) error {
+	rules, err := p.cfg.CompileRules()
+	if err != nil {
+		return err
+	}
+
+	p.rules = rules
+	return nil
+}
+
+// processLogs masks incoming logs.
+func (p *processor) processLogs(_ context.Context, ld plog.Logs) (plog.Logs, error) {
+	for i := 0; i < ld.ResourceLogs().Len(); i++ {
+		resource := ld.ResourceLogs().At(i)
+		resourceMaskFunc := p.createMaskFunc(resourceField)
```

Comment:
Do we need to do this every iteration for every log. Seems like we could create this once and save it off as it's not creating the function based on any specific log.

Reply:
I changed it so the processor has the resource and attributes range functions saved as fields. These fields are created when `Start` is called.